perf: optimize autocrafting calculation

SirYwell opened the pull request:
Also related to #1100, but while this change provides significant speedups, timeouts are still common when materials are missing.

One main observation is that we most likely don't touch most of the material in the system. Consequently, if we track the changes on top of the original system state, we can reduce copies to what the actual changes.

On top of that, we can avoid HashMap resizing by pre-sizing them better and we can also avoid lookups when copying as described in #956.

In my testings, even complex recipes get calculated when enough material is present. With material is missing, it is also a lot faster, but noticeably slower.

SirYwell commented on refinedstorage-resource-api/src/main/java/com/refinedmods/refinedstorage/api/resource/list/LazyCopyMutableResourceListImpl.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public MutableResourceList copy() {
+        return new LazyCopyMutableResourceListImpl(original, updates.copy(), new HashSet<>(pulled));
```
The `new HashSet<>(pulled)` call shows up now when profiling. I think it is fine, but there might be alternative approaches, e.g., creating a fresh HashSet instead and passing updated as a new original (i.e., build a deeper tree instead of the flat design right now).
That would add some more complexity.

It might also make sense to explore other Set implementations, (e.g., from fastutil) and see if that improves things further.

raoulvdberge replied:
I think we will always have some hot spot that will show up in the profiler, I think this is fine. We can introduce more complexity later.

However, I am interested in how fastutil performs on some hotspots (like `ResourceList`). But that is something I'd ideally like to measure/compare in another PR.

SirYwell commented on refinedstorage-resource-api/src/main/java/com/refinedmods/refinedstorage/api/resource/list/LazyCopyMutableResourceListImpl.java:
```diff
@@ -0,0 +1,112 @@
+ * An implementation of a {@link ResourceList} that lazily copies from an original list.
+ * Only the entries that are changed are copied.
+ */
+@API(status = API.Status.STABLE, since = "2.0.0-beta.11")
```
Not sure if this is the right

raoulvdberge replied:
It's right, but you can remove the `Impl` from the name. It's not needed here, as the name of the implementation is different than the interface.

raoulvdberge commented:
This is interesting. I will check this with some complex recipes on All The Mods.

In the meantime, I wrote this test (https://gist.github.com/raoulvdberge/029a5d0650a1f3f9d573534fb1ef3565) a while ago when trying to test some optimizations. At the time, I was trying to optimize autocrafting by introducing threading. Was a failed attempt, only made things slower.

But the test is still good, as it simulates complex recipes with lots of substitutions.
What do you think of it? How does it perform on your branch vs `develop`? Is it realistic?

SirYwell commented:
> This is interesting. I will check this with some complex recipes on All The Mods.
> 
> In the meantime, I wrote this test ([gist.github.com/raoulvdberge/029a5d0650a1f3f9d573534fb1ef3565](https://gist.github.com/raoulvdberge/029a5d0650a1f3f9d573534fb1ef3565)) a while ago when trying to test some optimizations. At the time, I was trying to optimize autocrafting by introducing threading. Was a failed attempt, only made things slower.
> 
> But the test is still good, as it simulates complex recipes with lots of substitutions. What do you think of it? How does it perform on your branch vs `develop`? Is it realistic?

Interesting test. My changes seem to perform worse there. It looks like sorting in the `IngredientState` constructor is the main contributor, as we now have 2 map lookups instead of 1 per list access (and there are 2 list accesses per comparison).
But the test is missing one central point this PR addresses: There are only 4 resources in the RootStorage. On our server, we currently have >14k resources there. That means this test doesn't really reflect the cost of copying all these resources properly.

But the slowdown in this test also comes down to one central pain point: The slow hashCode calculation of `ItemResource`. Maybe it makes sense to convert that record to a normal class and cache the hashCode once calculated (I'm not sure how often these are re-created, but with this kind of double-lookup I'd expect to see an improvement). I'll experiment with that.

raoulvdberge commented:
Would you say the test is too extreme if it comes down to crafting complexity, and is not extreme enough if it comes down to root storage size?

Then I wonder, what is the main reason why autocrafting recipes can't calculate in time on a typical Minecraft server: crafting complexity or root storage size?

I was thinking that it's crafting complexity, but maybe that is a wrong assumption, and my test is testing the wrong component, so it doesn't really matter that your code performs worse.

SirYwell committed:
perf: optimize ingredient state sorting

SirYwell commented:
Hm, at least in our scenario (crafting extra disk 1024k storage parts with the mentioned root storage size) the copying seems to be a far larger contributor. But I'm not sure if this is considered a complex recipe already.

One observation regarding the sorting: Even if the array is already sorted by coincidence, we need to get the amount of each resource at least once to verify it is sorted. That means we could pre-fetch the amount before sorting, and sort using that. It looks like we could reuse `CraftingState.ResourceState` there. From a quick test, it seems like we can get even better performance this way:

<table>
<tr>
 <td>Root Storage Size
 <td>develop
 <td>local changes
<tr>
 <td>4
 <td>239s
 <td>178s
<tr>
 <td>4 + 1000
 <td>322s
 <td>206s
</table>

I also noticed that `Pattern#hashCode` takes ~50s in every scenario above. Given a Pattern has a UUID, I wonder if the hashCode could be simplified to return the hashCode of the UUID only? This would be an improvement orthogonal to this change.

SirYwell commented:
Another observation, and -- if I didn't miss any problems -- potentially the most relevant optimization: `PatternLayout` currently has an Ingredient list. For recipes where one ingredient appears in multiple slots (e.g., all the storage part recipes, where you need 3 parts of the previous tier), this means the list potentially contains the same *resource* multiple times. But that means crafting Item A from 3xB calculates 3 subtrees, while crafting 3xB directly only calculates one tree. *Especially* in these typical recipes where multiple lower tiers are used to craft higher tiers, this means we run into exponential search complexity, when linear complexity would be enough.

As an example, instead of the following search tree:

```mermaid
flowchart TD
    A[4k Storage Part]
    A -->|1| B1
    A -->|1| B2
    A -->|1| B3
    A -->|1| B4
    A -->|1| B5
    A -->|1| B6
    A -->|1| B7
    A -->|1| B8
    A -->|1| B9
    B1[1k Storage Part]
    B2[1k Storage Part]
    B3[1k Storage Part]
    B4[Basic Processor]
    B5[Basic Processor]
    B6[Basic Processor]
    B7[Basic Processor]
    B8[Redstone]
    B9[Iron Ingot]
    B1 -->|1| C1[Silicon]
    B1 -->|1| C2[Silicon]
    B1 -->|1| C3[Silicon]
    B1 -->|1| C4[Silicon]
    B1 -->|1| C5[Glass]
    B1 -->|1| C6[Glass]
    B1 -->|1| C7[Glass]
    B1 -->|1| C8[Iron Ingot]
    B1 -->|1| C9[Redstone]
```
    
 we can simplify the search to
 
 ```mermaid
flowchart TD
    A[4k Storage Part]
    A -->|3| B1
    A -->|4| B4
    A -->|1| B8
    A -->|1| B9
    B1[1k Storage Part]
    B4[Basic Processor]
    B8[Redstone]
    B9[Iron Ingot]
    B1 -->|3 * 4| C1[Silicon]
    B1 -->|3 * 3| C5[Glass]
    B1 -->|3* 1| C8[Iron Ingot]
    B1 -->|3 * 1| C9[Redstone]
 ```
 In my tests, this results in basically instant results even for complex recipes and also when resources are missing.

 Is there anything I'm missing that would prevent this optimization?

SirYwell commented:
I pushed my changes to the IngredientState possibilities sorting and PatterLayout Ingredient deduplication as separate commits so you can take a look. Please let me know what you think.

raoulvdberge commented:
I love it. Bit busy with work this week, I'll try to take a look as soon as I can. Thanks a lot so far for all the investigation.

raoulvdberge commented:
In the meantime, please check the test failures and checkstyle problems.
I think there were also some mutation testing problems in an earlier commit (this means that some new tests are missing, or existing tests need more assertions.)

SirYwell commented:
Looking into the test failures:

- `PreviewTest`, `TreePreviewTest` (`shouldPrioritizeResourcesThatWeHaveMostOfInStorageForSingleRootPatternAndMultipleIngredients()`): My fault! The `compare` implementation was wrong, good that these tests exist :)
- The other tests seem all to fail for the same reason, there is indeed different behavior now:
	-  Previously, when crafting 3 crafting tables from `4 x oak planks, 1 x oak log, 1 x spruce log`, we would for each of the 4 planks needed for a crafting table try to find 3 resources.
		1. Use 3 of the 4 oak planks
		2. Use the last oak plank and craft 4 more (=> 2 oak planks in the internal storage)
		3. Use the 2 oak planks and craft 4 spruce planks (=> 3 spruce planks in the internal storage)
		4. Use the remaining 3 spruce planks
	- With my most recent change, we look once for 12 planks instead.
		- With crafting we get to 8 oak planks, but that's not enough => MISSING_RESOURCES
		- With crafting we get to 4 spruce planks, but that's not enough => MISSING_RESOURCES
	- This seems odd, not sure if this is the intended behavior. Especially as we had mixed planks before already! I think MISSING_RESOURCES should only be the result if all possibilities together cannot satisfy the needs. But I wonder if crafting should be preferred over using multiple different resources in that case. If not, we would need to track craftable items and try them after all available resources were checked/consumed.

SirYwell commented:
I updated the commits. Regarding the mutation tests, I reworked some stuff to allow predefined lists (needed for the lazy copy one, as it's impossible to add to the original list later). Please let me know what you think of the changes there.

The sorting is fixed now, but the crafting issue remains for now. I'll wait for your input there before proceeding with anything.

raoulvdberge commented:
Ah yes, it is important that we keep the existing behavior. It needs to process these ingredient by ingredient in order to make sure that mixed ingredients still work; this is what makes RS2 autocrafting so powerful.

SirYwell committed:
perf: optimize autocrafting calculation

SirYwell commented:
> Ah yes, it is important that we keep the existing behavior. It needs to process these ingredient by ingredient in order to make sure that mixed ingredients still work; this is what makes RS2 autocrafting so powerful.

That's unfortunate, I removed the last patch (but I still got it here https://gist.github.com/SirYwell/857e2b6766698ac9b4633fcdc97a2be0 in case it gets relevant again somehow). I wonder what parts of the behavior need to be kept 1:1 and to what extent the behavior could be changed while keeping RS2 powerful. Maybe there are different ways to get similar speedups without violating these constraints. 

During this deep-dive I also noticed that there are cases where it produces unexpected results:

1. Create a fuzzy pattern for sticks
2. Create a pattern for oak fence gate
3. Put 2 x oak planks and 2 x other planks into the system
4. Try crafting an oak fence gate. Instead of using the other planks for the sticks, 1 oak plank is used for the sticks. Then, only 1 is available anymore but 2 are needed.

I assume resolving that would need a DAG instead of a tree, with some kind of dynamic reallocation of resources. But that sounds even more computation-heavy (but maybe allows for deduplication?). I'll try to do more research on that.

raoulvdberge commented:
Okay, some more time now. Like I mentioned I wanted to do some testing myself too.

I did some testing on All The Mods 10 crafting the **1048m Storage Disk** (fuzzy mode "on", on all patterns) and seeing promising results.

Analysing the table, it seems like this PR resolves the time-out problem for large crafts with no missing ingredients AND lots of unique items (before, it would time-out even if we had enough ingredients).

| Branch    | Missing Ingredients | Unique Items | Result     |
|-----------|---------------------|--------------|------------|
| develop   | Yes                 | 0          | Time-out  |
| develop   | No                  | 10          | No time-out  |
| develop   | Yes                 | 55633           |   (Didn't test, assuming this will also time-out given that missing ingredients with 0 unique items already times out) |
| develop   | No                  | 55633           | Time-out  |
| Your PR   | Yes                 | 0          |   Time-out  |
| Your PR   | No                  | 10          |  No time-out |
| Your PR   | Yes                 | 55633           | (Didn't test, assuming this will also give a time-out given that missing ingredients with 0 unique items already times out) |
| Your PR   | No                  | 55633           | No time-out  |

(This is consistent with your findings mentioned in the PR description)

> - 10 unique items: all the base ingredients necessary for the 1048m Storage Disk
> - 55633 unique items: all the items in the item registry EXCLUDING:
>   - Items from Extra Storage
>   - Items from Extra Disks
>   - Items from Refined Storage
>   - Glass (to ensure it uses the smelting recipe from Sand)
>   - Iron (to ensure it uses the smelting recipe from Iron Ingots)

My conclusion:

- No missing ingredients, and few unique items: never caused an issue
- Missing ingredients, few unique items: an issue on `develop` but also on this PR
- No missing ingredients, and lots of unique items: causes an issue on `develop`, and fixed by this PR
- Missing ingredients, and lots of unique items: an issue on `develop` but also on this PR

And a conclusion regarding the performance of your PR on my complex crafting test (please challenge me if incorrect):

- My complex crafting test is NOT realistic compared to complex recipes in Minecraft.
  - Why? The test completes in 30 sec., while complex crafts in Minecraft calculate in around 2 sec.
  - This indicates to me that the depth/complexity of the crafted resource in the test is too high.
- However, I do think it is a good benchmark for the autocrafting code.
- Right now it isolates crafting complexity. To be even more realistic, it would need to have a bigger root storage to also take the copying cost into account.

raoulvdberge commented:
Some questions I have for you:

- Could you rebase this PR on the latest `develop`? I'll merge this PR after that.
- Does this PR fix issue #956? (I'm assuming yes)
- Could you look into using fastutil in some hotspots like `ResourceList` and tell me if there is a measurable difference (in a separate PR)?
- Like you mentioned, how does improving the hashcode calculation of `ItemResource` and `Pattern` affect the performance?
- How can we avoid time-outs when there are missing ingredients but keep the existing behavior (or maybe a loosened variant of that)?
  - How far can we get with optimization fixes like this PR, without overhauling the entire autocrafting logic? 

Thanks for all the research you did so far. I'm really happy someone wants to help out with performance!

SirYwell commented:
> * Could you rebase this PR on the latest `develop`? I'll merge this PR after that.

Sure, will do so later.

> * Does this PR fix issue #956?

At least to some extent. If ingredients are missing, I'd expect the problem to be less extreme but still existing. But now the performance isn't directly coupled to the number of unique items anymore, so "cheap" recipes with high numbers of unique items should definitely see improvements. But more below regarding hashing.

> * Could you look into using fastutil in some hotspots like `ResourceList` and tell me if there is a measurable difference (in a separate PR)?

I can, but I'd first look into hash code calculation, that seems to be a bigger contributor right now.

> * Like you mentioned, how does improving the hashcode calculation of `ItemResource` and `Pattern` affect the performance?

I'd need to take new measurements there, but I have some motivating examples for `ItemResource`:
- Example of autocrafting using an exporter: https://spark.lucko.me/S70TBsnvL2 (look for `ItemResource.hashCode` in the flat view)
- Example of interactions via the Advanced Peripherals RS Bridge: https://spark.lucko.me/zI58wF9cOW?hl=1940 (though I think this also is sub-optimal usage of the API, as `getAll` would probably enough?)
- Example of a client lag when interacting with JEI: https://spark.lucko.me/j2qV5byfu9?hl=2329
- Example of `PreviewBuilder.copy` (I disabled the timeout there, the whole preview calculation was 16 seconds): 
<img width="1451" height="161" alt="preview-builder-copy" src="https://github.com/user-attachments/assets/ee759c84-9abb-4711-8a95-d09ba552ba40" />

- In the same example, adding and removing the pattern to `activePatterns` in `CraftingTree` takes around 900ms. This is the only case where I saw it being somewhat relevant. So while it's not that much in sum, basically all of it is spent on hashing `PatternLayout`. If hashing the UUID is enough, that would be an extremely simple improvement.

> * How can we avoid time-outs when there are missing ingredients but keep the existing behavior (or maybe a loosened variant of that)?
> 
>     * How far can we get with optimization fixes like this PR, without overhauling the entire autocrafting logic?

Keeping the existing behavior exactly like that makes improvements that aren't just very local rather difficult. While it might be possible to squeeze out a bit more performance with micro-optimizations (e.g. replacing streams with loops), we're getting closer to the point where the limitation lies more in the algorithm than the implementation.

The change I experimented with could maybe get close to the current behavior with more adjustments. I'm not sure how much effort that would be though. Recall https://github.com/refinedmods/refinedstorage2/pull/1129#issuecomment-3220518111, a possible change would need to somehow keep track of partial results (i.e., first try to satisfy the requirement using *multiple* resources if multiple variants are allowed, not just one).

I'm not sure how well that would work with fuzzy processing patterns, when e.g., 2 stacks of logs are required: With this change, we would allow stuff like 1x oak log + 127 spruce log. Would it be a requirement to have 2 stacks of 1 material each in this case? 
Maybe there could be 2 paths, and deduplication just doesn't happen for processing patterns then.

SirYwell commented:
> * And a conclusion regarding the performance of your PR on my complex crafting test (please challenge me if incorrect):
> 
>     * My complex crafting test is NOT realistic compared to complex recipes in Minecraft.
>       
>       * Why? The test completes in 30 sec., while complex crafts in Minecraft calculate in around 2 sec.
>       * This indicates to me that the depth/complexity of the crafted resource in the test is too high.
> 
>     * However, I do think it is a good benchmark for the autocrafting code.
> 
>     * Right now it isolates crafting complexity. To be even more realistic, it would need to have a bigger root storage to also take the copying cost into account.

That's correct so far. I'm not sure how well that is covered right now, but having the same ingredient multiple times (and the same for patterns for that ingredient) blows up the search when resources are missing, because (more or less) the same search is done multiple times.

raoulvdberge commented:
Regarding this message https://github.com/refinedmods/refinedstorage2/pull/1129#issuecomment-3220518111 I'm not sure I can follow this statement:

> This seems odd, not sure if this is the intended behavior. Especially as we had mixed planks before already! I think MISSING_RESOURCES should only be the result if all possibilities together cannot satisfy the needs. But I wonder if crafting should be preferred over using multiple different resources in that case. If not, we would need to track craftable items and try them after all available resources were checked/consumed.

MISSING_RESOURCES is only returned if none of the ingredients work, currently.

Yes, if you combine ingredients then it will no longer differentiate between slots and MISSING_RESOURCES could be a misleading result.

Crafting should not be preferred over using multiple resources. But how would tracking craftable items solve this issue, if you end up combining all the inputs?

> a possible change would need to somehow keep track of partial results (i.e., first try to satisfy the requirement using multiple resources if multiple variants are allowed, not just one).

I don't understand this. Currently, it's checking multiple resources already.

> I'm not sure how well that would work with fuzzy processing patterns, when e.g., 2 stacks of logs are required: With this change, we would allow stuff like 1x oak log + 127 spruce log. Would it be a requirement to have 2 stacks of 1 material each in this case?

It's a good question. My assumption is that mixing within 1 slot will give trouble.

raoulvdberge commented:
I'm struggling to wrap my head around https://github.com/refinedmods/refinedstorage2/pull/1129/commits/d056b19f50fa80ebb70e66641bad26b30fe1dd8d

Won't calculating the resources in storage before give problems?
Let's say we require an ingredient with possibilities [A,B].

We first calculate A, we do a deep subcraft and we mutate the internal storage.
We run out of items for A, so we cycle the ingredient and try B.

In the old code, this would calculate the resources in storage **at this point**.

However, in the new code, the amounts are pre-calculated. 

But wouldn't this sorting be incorrect, not accounting of the changes being made for A?

Is there something that I'm missing?

SirYwell commented:
> Regarding this message [#1129 (comment)](https://github.com/refinedmods/refinedstorage2/pull/1129#issuecomment-3220518111) I'm not sure I can follow this statement:

Sorry for the confusion, this is indeed about the case when multiple ingredients that allow the same resources are combined.

> Yes, if you combine ingredients then it will no longer differentiate between slots and MISSING_RESOURCES could be a misleading result.

> It's a good question. My assumption is that mixing within 1 slot will give trouble.

I think that's the underlying problem. But it *might* be safe for crafting patterns. In that case, fixing the calculation based on my deduplication changes would need two code paths, one for crafting (="it is fine to use different materials for this ingredient") vs processing (="one slot must only use one material"). That might get quite complex.

> Crafting should not be preferred over using multiple resources. But how would tracking craftable items solve this issue, if you end up combining all the inputs?

I'd need to investigate, I might be wrong there. But if I want to craft 3x crafting table, even currently, we look for 3x planks for each slot. And for each slot, only 1 material is allowed (iirc). In that case, we craft planks instead of trying to use planks of different types for that slot for the different crafting tables. With ingredient deduplication, due to the higher amount needed, we also try to craft more (and if we can't, we fail). **If** using multiple materials for one (deduplicated) ingredient is allowed, it would make sense to first use the available materials (and mix them). And if that's not enough, try crafting more of any of them. "Tracking" in some sense would be a heuristic for which material to start with. I'm not sure if that makes any sense though.

I hope that makes it a bit clearer. There are many difficult details involved, making it really hard to explain, sorry.

SirYwell commented:
> I'm struggling to wrap my head around [d056b19](https://github.com/refinedmods/refinedstorage2/commit/d056b19f50fa80ebb70e66641bad26b30fe1dd8d)
> 
> Won't calculating the resources in storage before give problems? Let's say we require an ingredient with possibilities [A,B].
> 
> We first calculate A, we do a deep subcraft and we mutate the internal storage. We run out of items for A, so we cycle the ingredient and try B.
> 
> In the old code, this would calculate the resources in storage **at this point**.
> 
> However, in the new code, the amounts are pre-calculated.
> 
> But wouldn't this sorting be incorrect, not accounting of the changes being made for A?
> 
> Is there something that I'm missing?

Hm, the sorting shouldn't be affected:
It happens at the same time at the same place. The only difference is that we look up the amount directly before sorting, instead of during sorting. But in between, there is nothing modifying the amount, right?

However, regarding cycling: If the internal storage changes the amount of B due to A, then B would give an outdated result. I'm not sure if that can happen? Similarly, if A runs into crafting something that requires B, then we'd modify the storage to have less of B. Could that happen?

But if such effects should be reflected in the array order, then you would need to re-sort the tail of the array every time I think?

raoulvdberge commented:
I was trying to write a test to prove that it could break, but I couldn't. Indeed, it can never happen, for this reason:

> But wouldn't this sorting be incorrect, not accounting of the changes being made for A?

We don't care about the changes for A, as the subcraft for A has failed (there were not enough resources to craft A).
So the intermediates for A are dismissed and it will try B.
The system does not handle partial crafts (doing that as well would only make things slower), the full subcraft has to succeed.
There is nothing that A could do that could affect the sorting or inStorage counts later.

Given that, the sorting can perfectly be predetermined.

And given that, I'll merge this PR. Thanks a lot!
